README.md
# jQuery seSlider #

jquery.seSlider is a lightweight slider plugin for jQuery.
It allows you to create classical slider with next and previous button or slideshow that can be linked to a sound.

Requirements
-----
[jQuery](http://www.jquery.com) (>= 1.7)

Usage
-----

### Step 1: Call the JS lib ###
First, include the jQuery library (by downloading it or calling it from CDN).  
Next, download the package from this site and link the slider Javascript file.

```html
<!-- jQuery library (served from Google CDN) -->
<script src="//ajax.googleapis.com/ajax/libs/jquery/2.1.4/jquery.min.js"></script>
<!-- slider Javascript file -->
<script src="/js/jquery.seslider.min.js"></script>
```

### Step 2: Create HTML markup ###
Create a `<div class="slider">` tag, with a `<ul>` list and a `<li>` for each slide. Slides can contain images, video, or any HTML content.

```html
<div class="slider">
    <ul>
        <li><img src="img/img01.jpg" alt="" /></li>
        <li><img src="img/img02.jpg" alt="" /></li>
        <li><img src="img/img03.jpg" alt="" /></li>
        ...
    </ul>
</div>
```

### Step 3: Add the following CSS to your code. ###
You will have to edit the width and height properties to adapt the slider to your needs. 

```css
.slider {
    width: 800px; /* can be changed following your needs */
    height: 600px; /* can be changed following your needs */
    overflow: hidden;
}

.slider ul {
    margin: 0;
    padding: 0;
    position: relative;
    list-style: none;
}

.slider ul li {
    width: 800px; /* can be changed following your needs */
    float: left;
}
```

### Step 4: Call the plugin ###

```javascript
$(document).ready(function() {
    $('.slider').seSlider();
});
```

Options
-----
- **nextBtn**  
The next button selector.  
*default: '.sliderNextBtn'*

- **prevBtn**  
The previous button selector.  
*default: '.sliderPrevBtn'*

- **playPauseBtn**  
The play/pause button selector.  
*default: '.sliderPlayBtn'*

- **resetBtn**  
The reset button selector.  
*default: null*

- **preventReversedCycle**  
Allow or prevent changing slide if you are on the first slide on click on previous button.  
*default: false*

- **progressBar**  
The progress bar selector. If the value is different to null the plugin will automaticly update the progressBar.  
*default: null*

- **changeCallback**  
Function called when the change slide animation is triggered.  
*default: null*  
*Optional parameters: currentSlideIndex, toSlideIndex, direction*

- **afterChangeCallback**  
Function called after the change slide animation is complete.  
*default: null*  
*Optional parameters: newSlideIndex*

- **playCallback**  
Function called when the slideShow starts.  
*default: null*

- **pauseCallback**  
Function called when the slideShow is set to pause.  
*default: null*

- **endCallback**  
Function called when the slideShow is ended.  
*default: null*

- **resetCallback**  
Function called when the slideShow is reseted.  
*default: null*

- **slideshowSoundTrack**  
The audio tag selector that runs with the slideshow.  
*default: null*

- **slideshowSteps**  
The array containing the times of each slide.  
If the value is set to null the slideshow will change to the next slide after each `slideshowIntervalTime`.  
*default: null*

- **slideshowIntervalTime**  
If `slideshowSteps` is null: The interval time between each slide change.  
If `slideshowSteps` is not null: The interval time between each step checking.  
*default: 1000*

- **transitionSpeed**  
The animation's transition speed between 2 list elements.  
*default: 200*

Progress bar
-----

Due to a desire of compatibility the plugin is designed to interact with a progress bar composed by a classical div tag

```html
<div class="progressBar"><div></div></div>
```

```css
.progressBar {
    width: 800px; /* can be changed following your needs */
    height: 15px; /* can be changed following your needs */
    box-shadow: 0 2px 5px #aaaaaa inset; /* can be changed following your needs */
    background: #eeeeee; /* can be changed following your needs */
}

.progressBar div {
    width: 0;
    height: 100%;
    background: #ee4444; /* can be changed following your needs */
}
```


CHANGELOG.md
# 1.2.1 / 2016-06-11
* Bug fixes

# 1.2.0 / 2015-08-20
* Add resetCallback
* Prevent error if the slideshowSteps length is different to the number of li tags
* Bug fixes

# 1.1.0 / 2015-08-19
* Add changeCallback and afterChangeCallback
* Global optimization

# 1.0.0 / 2015-07-30
* First release


gulpfile.js
var gulp = require('gulp'),
    rimraf = require('rimraf'),
    jshint = require('gulp-jshint'),
    uglify = require('gulp-uglify'),
    rename = require('gulp-rename'),
    insert = require('gulp-insert'),
    p = require('./package.json')
    ;

//Folders and annotation
var src = 'src/',
    dest = 'dist/',
    annotation = '/* ' + p.name + ' v' + p.version + ' | Copyright (c) ' + new Date().getFullYear() + ' ' + p.homepage + ' | ' + p.license + ' license */\n';

gulp.task('clean', function(cb) {
    rimraf(dest, cb);
});

gulp.task('scripts', ['clean'], function() {
    return gulp.src(src + '*.js')
        .pipe(jshint())
        .pipe(jshint.reporter())
        .pipe(uglify())
        .pipe(rename('jquery.seslider.min.js'))
        .pipe(insert.prepend(annotation))
        .pipe(gulp.dest(dest))
        ;
});

gulp.task('default', ['scripts'], function() {

});


src/jquery.seslider.js
(function($) {
    $.fn.seSlider = function(options) {
        var defaults = {
            nextBtn: '.sliderNextBtn',
            prevBtn: '.sliderPrevBtn',
            playPauseBtn: '.sliderPlayBtn',
            resetBtn: '.sliderResetBtn',
            preventReversedCycle: false,
            progressBar: null,
            changeCallback: null,
            afterChangeCallback: null,
            playCallback: null,
            pauseCallback: null,
            endCallback: null,
            resetCallback: null,
            autoPlay: false,
            slideshowSoundTrack: null,
            slideshowSteps: null,
            slideshowIntervalTime: 1000,
            transitionSpeed: 200
        };
        var params = $.extend(defaults, options),
            changeCallbacks = $.Callbacks(),
            afterChangeCallbacks = $.Callbacks();

        if(typeof params.changeCallback === 'function')
            changeCallbacks.add(params.changeCallback);

        if(typeof params.afterChangeCallback === 'function')
            afterChangeCallbacks.add(params.afterChangeCallback);

        function moveLeft(slider, slideshow) {
            if(params.preventReversedCycle === true && slideshow.currStep <= 0)
                return;

            changeCallbacks.fire(slideshow.currStep, slideshow.currStep - 1, 'left');

            slider.animate({
                left: + slider.children('li').width()
            }, params.transitionSpeed, function() {
                slideshow.currStep--;
                slider.find('li').last().prependTo(slider);
                slider.css('left', '');

                if(slideshow.currStep < 0)
                {
                    if(slideshow.interval !== false)
                        stopSlideShow(slider, slideshow, true, false);

                    slideshow.currStep = slideshow.maxStep;
                }

                afterChangeCallbacks.fire(slideshow.currStep);
                updateElapsedTime(slideshow);
                updateProgressBar(slideshow.elapsedTime / slideshow.maxTime * 100);
                updateSoundTrackTime(slideshow.elapsedTime / 1000);
            });
        }

        function moveRight(slider, slideshow, updateComponents) {
            changeCallbacks.fire(slideshow.currStep, slideshow.currStep + 1, 'right');

            slider.animate({
                left: - slider.children('li').width()
            }, params.transitionSpeed, function() {
                slideshow.currStep++;
                slider.find('li').first().appendTo(slider);
                slider.css('left', '');

                if(slideshow.currStep > slideshow.maxStep)
                {
                    slideshow.currStep = 0;
                    if(slideshow.interval !== false)
                    {
                        stopSlideShow(slider, slideshow, true, false);
                        updateComponents = false;
                    }
                }

                afterChangeCallbacks.fire(slideshow.currStep);
                if(updateComponents === true)
                {
                    updateElapsedTime(slideshow);
                    updateProgressBar(slideshow.elapsedTime / slideshow.maxTime * 100);
                    updateSoundTrackTime(slideshow.elapsedTime / 1000);
                }
            });
        }

        function playSlideShow(slider, slideshow)
        {
            slideshow.interval = setInterval(function() {
                slideshow.elapsedTime += params.slideshowIntervalTime;
                updateProgressBar(slideshow.elapsedTime / slideshow.maxTime * 100);

                if(params.slideshowSteps === null || params.slideshowSteps[slideshow.currStep] <= slideshow.elapsedTime)
                    moveRight(slider, slideshow, false);
            }, params.slideshowIntervalTime);

            if(params.slideshowSoundTrack !== null)
                $(params.slideshowSoundTrack).get(0).play();

            if(typeof params.playCallback === 'function')
                params.playCallback();
        }

        function stopSlideShow(slider, slideshow, endReached, reset)
        {
            if(slideshow.interval !== false)
            {
                clearInterval(slideshow.interval);
                slideshow.interval = false;

                if(params.slideshowSoundTrack !== null)
                    $(params.slideshowSoundTrack).get(0).pause();
            }

            if(endReached === true || reset === true)
            {
                slider.finish();
                if(reset === true && slideshow.currStep !== 0)
                    slider.find('li').slice((slideshow.maxStep + 1) - slideshow.currStep).prependTo(slider);

                slideshow.currStep = 0;
                slideshow.elapsedTime = 0;
                updateProgressBar(0);
                updateSoundTrackTime(0);

                if(endReached === true && typeof params.endCallback === 'function')
                    params.endCallback();
                else if(reset === true && typeof params.resetCallback === 'function')
                    params.resetCallback();
            }
            else if(typeof params.pauseCallback === 'function')
                params.pauseCallback();
        }

        function updateElapsedTime(slideshow)
        {
            if(params.slideshowSteps === null)
                slideshow.elapsedTime = params.slideshowIntervalTime * slideshow.currStep;
            else
                slideshow.elapsedTime = params.slideshowSteps[slideshow.currStep - 1] || 0;
        }

        function updateProgressBar(size)
        {
            if(params.progressBar !== null)
                $(params.progressBar).children('div').css('width', size + '%');
        }

        function updateSoundTrackTime(time)
        {
            if(params.slideshowSoundTrack !== null)
                $(params.slideshowSoundTrack).get(0).currentTime = time;
        }

        return this.each(function() {
            var obj = $(this).children('ul');
            var slideWidth = obj.children('li').width(),
                slideshow = {
                interval: false,
                maxStep: obj.children('li').length - 1,
                currStep: 0,
                elapsedTime: 0
            };

            if(slideshow.maxStep < 1)
                return true;

            if(params.slideshowSteps !== null && params.slideshowSteps.length != slideshow.maxStep + 1)
                throw new TypeError('The slideshowSteps parameter must have a length equal to the number of li tags in the slider.');

            slideshow.maxTime = params.slideshowSteps === null ? params.slideshowIntervalTime * (slideshow.maxStep + 1) : params.slideshowSteps[slideshow.maxStep];

            obj.children('li').last().prependTo(obj);
            if(obj.children('li').length === 2)
            {
                var first = obj.children('li').first().clone(),
                    last = obj.children('li').last().clone();
                obj.append(first).append(last);
            }

            obj.css({width: slideWidth * (slideshow.maxStep + 1), marginLeft: - slideWidth});

            if(params.autoPlay === true)
                playSlideShow(obj, slideshow);

            //Next button
            $(params.nextBtn).on('touchstart click', function(e) {
                e.preventDefault();
                if(obj.css('left') !== 'auto' && obj.css('left') !== '0px')
                    return;

                moveRight(obj, slideshow, true);
            });

            //Prev button
            $(params.prevBtn).on('touchstart click', function(e) {
                e.preventDefault();
                if(obj.css('left') !== 'auto' && obj.css('left') !== '0px')
                    return;

                moveLeft(obj, slideshow);
            });

            //Reset button
            $(params.resetBtn).on('touchstart click', function(e) {
                e.preventDefault();
                stopSlideShow(obj, slideshow, false, true);
            });

            //SlideShow play
            $(params.playPauseBtn).on('touchstart click', function(e) {
                e.preventDefault();
                if(slideshow.interval === false)
                    playSlideShow(obj, slideshow);
                else
                    stopSlideShow(obj, slideshow, false, false);
            });
        });
    };
})(jQuery);
